homeassistant/components/mysensors.py

```diff
@@ -0,0 +1,219 @@
+
+
+New features:
+
+New MySensors component.
+Updated MySensors Sensor platform.
+New MySensors Switch platform.
+Multiple gateways are now supported.
+
+Configuration.yaml:
+
+mysensors:
+  port:
+    - '/dev/ttyUSB0'
+    - '/dev/ttyACM1'
+  debug: true
+  persistence: true
+  persistence_file:
+    - 'path/to/.homeassistant/mysensors.json'
+    - 'path/to/.homeassistant/mysensors2.json'
+  version: '1.5'
+
+sensor:
+  platform: mysensors
+
+switch:
+  platform: mysensors
+"""
+import logging
+
+from homeassistant.helpers import (validate_config)
+
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START,
+    EVENT_HOMEASSISTANT_STOP,
+    TEMP_CELCIUS)
+
+CONF_PORT = 'port'
+CONF_DEBUG = 'debug'
+CONF_PERSISTENCE = 'persistence'
+CONF_PERSISTENCE_FILE = 'persistence_file'
+CONF_VERSION = 'version'
+
+DOMAIN = 'mysensors'
+DEPENDENCIES = []
+REQUIREMENTS = [
+    'https://github.com/theolind/pymysensors/archive/'
+    '2aa8f32908e8c5bb3e5c77c5851db778f8635792.zip#pymysensors==0.3']
+_LOGGER = logging.getLogger(__name__)
+ATTR_PORT = 'port'
+ATTR_DEVICES = 'devices'
+ATTR_NODE_ID = 'node_id'
+ATTR_CHILD_ID = 'child_id'
+ATTR_UPDATE_TYPE = 'update_type'
+
+IS_METRIC = None
+CONST = None
+GATEWAYS = None
+EVENT_MYSENSORS_NODE_UPDATE = 'MYSENSORS_NODE_UPDATE'
+
+
+def setup(hass, config):  # noqa
+    """ Setup the MySensors component. """
+    # pylint:disable=no-name-in-module
+    import mysensors.mysensors as mysensors
+
+    if not validate_config(config,
+                           {DOMAIN: [CONF_PORT]},
+                           _LOGGER):
+        return False
+
+    version = config[DOMAIN].get(CONF_VERSION, '1.4')
+
+    global CONST
+    if version == '1.4':
+        import mysensors.const_14 as const
+        CONST = const
+    elif version == '1.5':
+        import mysensors.const_15 as const
+        CONST = const
+    else:
+        import mysensors.const_14 as const
+        CONST = const
+
+    # Just assume celcius means that the user wants metric for now.
+    # It may make more sense to make this a global config option in the future.
+    global IS_METRIC
+    IS_METRIC = (hass.config.temperature_unit == TEMP_CELCIUS)
+
+    def callback_generator(port, devices):
+        """Return a new callback function. Run once per gateway setup."""
+        def node_update(update_type, nid):
+            """Callback for node updates from the MySensors gateway."""
+            _LOGGER.info('update %s: node %s', update_type, nid)
+
+            hass.bus.fire(EVENT_MYSENSORS_NODE_UPDATE, {
+                ATTR_PORT: port,
+                ATTR_DEVICES: devices,
+                ATTR_UPDATE_TYPE: update_type,
+                ATTR_NODE_ID: nid
+            })
+            return
+        return node_update
+
+    def setup_gateway(port, persistence, persistence_file):
+        """Return gateway after setup of the gateway."""
+        devices = {}    # keep track of devices added to HA
```

Comment:
Devices is just passed around but it does not seem to be used. This is good because events sent over the eventbus need to be fully serializable. Device objects are not.